### PR TITLE
Adds support for tmuxinator start --append

### PR DESCRIPTION
**Pull request by flovilmart**
Closes: #265

### CODE REVIEW

**Comment by ethagnawl** [comment]
Thanks for this, @flovilmart! I'll try to give it a proper review before the end of the week.

**Comment by DestyNova** [comment]
Very nice. The only confusing bit is that it names the new window `<session-name>:<window-name>` -- i.e. if I append a window called `muffin` from a file containing the line `name: mountain`, then the new window will be called `mountain:muffin`. If I don't specify the `name` parameter in the Yaml file, then the window will be called `:muffin`.

I see there is some logic to control this behaviour, but I don't fully understand it. Is there an extra flag or configuration value needed to just have the window be called `muffin`?

**Comment by flovilmart** [comment]
@DestyNova that would be a possible change, if the name is not set with the tmux template I could probably remove the separator. This was intended to make it easy to see which windows were appended.

**Comment by DestyNova** [comment]
Yes I see the motivation for it and it makes sense -- in my case though, I expect to end up with a bunch of files that each describe one window with a couple of panes. Since I use windows and panes where other people might use sessions and windows, I don't need that namespacing. Maybe the behaviour could be specified with an optional flag (e.g. if a simple name is default, then add `-p` for `prefix-with-session-name`, or if the namespaced window name is default, then `-s` for `simple-window-name` or something)?

**Comment by flovilmart** [comment]
I’m not sure I want to spend additional time adding more configuration flags to this PR. If the naming of the windows is problematic i’ll Remove it altogether and restore the original behavior instead of prefixing with the screen name.

**Comment by DestyNova** [comment]
I like the PR either way and am really happy someone has finally done this. Rather than sidetrack the PR, I'll try to do a separate one adding such a flag afterwards. Thanks for doing this :)

**Comment by ethagnawl** [comment]
@flovilmart This is looking pretty good!

I do have a few questions/thoughts/requests/etc. for you:

1.) Would you mind adding an entry to the "Other Commands" section of README.md in your PR which documents how this feature and its various CLI options are intended to be used?

The documentation will be useful for users going forward _and_ for people reviewing this PR. I wasn't following #265 closely and it took me a few minutes to figure out how this feature was intended to work.

2.) A missing/malformed `current_session: ...` seems to result in tmuxinator wantonly appending windows to whatever the most recently created tmux session is, which seems undesirable. Is this by design? Is there a way you can sniff out whether or not the `current_session` exists and either fail or start a new session if it doesn't? (I'll revisit the code myself, but I'm on my way out for the day.)

3.) Could you replace `puts` with `say` in Cli.rb?

4.) Perhaps the "Creating new session ..." message should contain the session name?

5.) We will need some tests around this new behavior. Did you ever get a chance to take a pass at writing any? I know you said you're not used to Ruby, so I can appreciate how that might be a bit daunting and it's awesome that you were able to put this together. (I don't think you're giving yourself enough credit!) If not, that's totally fine and I can take a pass at writing some and submit another PR to your branch.

**Comment by flovilmart** [comment]
I Didn’t forget about this one, but my todo list keeps pushing it down. I hope to have more time soon

**Comment by raiyuza** [comment]
@flovilmart Any ETA, we in 2019 already :)

**Comment by flovilmart** [comment]
Well, yea, I need to but my head back into it :)

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by icole** [comment]
Any issue with getting this merged now that the merge conflicts are resolved?

**Comment by flovilmart** [comment]
@icole I need to work on the few tests and documentation. I haven't found a good time to do it. I'll try this weekend.

**Comment by icole** [comment]
@flovilmart awesome, thanks for the response. If you could use any help with anything let me know. Would love to see this go out.

**Comment by flovilmart** [comment]
Thanks @icole I'll do my best to address all the concerns and add the relevant tests. I'd love to get done with this one. Sorry to have dropped the ball.

**Comment by flovilmart** [comment]
@icole Sorry for the late response. I changed a bit the logic so it uses `mux start --append [project]` instead of `mux append`. This simplifies a bit the logic.

I'll add some tests around the `in_current_session` logic to make sure this is readable easily and maintainable as well.

**Comment by icole** [comment]
@flovilmart thanks so much! could we pull this in after you add those tests? I am dying to get this in since the version this is based off is older than what our projects Gemfile requires.

**Comment by betoharres** [comment]
I was trying to figure out if tmuxinator had this feature; well, hope it gets merged!

**Comment by flovilmart** [comment]
I completely forgot I had this one open, thanks for the ping 😅! I’ll do my best.

**Comment by protex** [comment]
Any chance this gets merged? It would be super useful.

**Comment by flovilmart** [comment]
@icole what what left for update? I wanna go ahead with this. Can you pin point in the code the different changes required for the merge?

**Comment by ethagnawl** [comment]
From what I recall, it's a minor Code Climate nitpick. Unfortunately, the build is old enough that it's been dropped by CC. Could you try pushing a new commit in order to trigger a new build, @flovilmart?

**Comment by flovilmart** [comment]
@ethagnawl the problem is that the `start` method now is too complex, not sure how to simplify it given it's already quite simple.

**Comment by ethagnawl** [comment]
I haven't used Code Climate in a long time and don't remember how some of their metrics work. I'd say maybe removing the `project` variable and nesting those method calls would work, but the more likely/correct solutions would be extracting those conditionals or moving the has creation into its own method.

Does anyone else have any thoughts about this or the PR in general? CC/ @tmuxinator/tmuxinator

**Comment by flovilmart** [comment]
@ethagnawl , finally, codeclimate passed! 🎉

**Comment by ethagnawl** [comment]
Okay, that's great. I'll have to find some time to review this again. 

Sorry for the bother, but does anyone else have any thoughts? /CC @tmuxinator/tmuxinator

**Comment by flovilmart** [comment]
thanks @ethagnawl !

File: lib/tmuxinator/config.rb [hunk 1]
```diff
@@ -25,6 +25,10 @@ def home?
         File.directory?(home)
       end
 
+      def options_file
+        directory + ".yml"
```
Comment (ethagnawl) [inline lib/tmuxinator/config.rb:29]: The "yaml" extension is also supported. I tried using a project config file with this extension and everything worked as expected. Does that imply that this method isn't being used or am I overlooking something?

File: lib/tmuxinator/project.rb [hunk 2]
```diff
@@ -208,7 +217,12 @@ def tmux_options
     end
 
     def base_index
-      get_base_index.to_i
+      base = 0
+      if in_current_session
+        # Get the last window index + 1
+        base = 1 + `tmux list-windows | awk '{print $1}'`.split.last.to_i
```
Comment (ethagnawl) [inline lib/tmuxinator/project.rb:223]: I think this awk logic should either be handled in Ruby code or we should add a method to the Doctor class which verifies that awk exists on the user's system.

File: lib/tmuxinator/cli.rb [hunk 3]
```diff
@@ -226,6 +227,44 @@ def show_continuation_prompt
...
         Kernel.exec(project.kill)
       end
+
+      def tmux_session_exists
+        ENV["TMUX"] ? true : false
+      end
+
+      def should_append?(options)
+        options[:append] == true ||
+          (options[:append] != false && Tmuxinator::Config.options[:append])
```
Comment (ethagnawl) [inline lib/tmuxinator/cli.rb:237]: Could you walk me through the use case for the right side of this or?

**Comment by kurktchiev** [comment]
any chance we can get this merged any time soon? it would really boost productivity with tmux/tmuxinator

File: lib/tmuxinator/cli.rb (hunk 3 above)
Comment (flovilmart) [reply lib/tmuxinator/cli.rb:237]: the idea is that can set append in the `Tmuxinator::Config.options` so all sessions are appending by default.
if append is not provided by the `options` (case `!= false`) we check that it is set in the global Config.

File: lib/tmuxinator/config.rb (hunk 1 above)
Comment (flovilmart) [reply lib/tmuxinator/config.rb:29]: what do you suggest?

File: lib/tmuxinator/project.rb (hunk 2 above)
Comment (flovilmart) [reply lib/tmuxinator/project.rb:223]: fair, let's me see what I can do

**Comment by cibinmathew** [comment]
@ethagnawl @flovilmart @icole Thanks for this work.
Can we get this merged or may be list down the blocking issues and open another to track the enhancements. We could pitch in to get this to a closure.

File: lib/tmuxinator/config.rb (hunk 1 above)
Comment (ethagnawl) [reply lib/tmuxinator/config.rb:29]: Either this method needs to support both extensions or, if I'm correct in that it's not being used, it should be removed.

**Comment by ethagnawl** [comment]
> @ethagnawl @flovilmart @icole Thanks for this work.
> Can we get this merged or may be list down the blocking issues and open another to track the enhancements. We could pitch in to get this to a closure.

There are still some issues (noted above) which need to be worked through. (If you feel like running with what's been done and @flovilmart doesn't mind, feel free to pick them up.) It'd also be good to get feedback from others in @tmuxinator/tmuxinator.

**Comment by flovilmart** [comment]
@ethagnawl let me get back on it

File: lib/tmuxinator/config.rb [hunk 4]
```diff
@@ -56,6 +60,15 @@ def default
...
       end
 
+      def options
+        yaml = begin
+          YAML.safe_load(File.read(options_file), [], [], true)
+        rescue SyntaxError, StandardError
+          return {}
+        end
+        yaml ? yaml : {}
+      end
```
Comment (noahfrederick) [inline lib/tmuxinator/config.rb:70]: I believe this is equivalent:

```suggestion
      def options
        YAML.safe_load(File.read(options_file), [], [], true) || {}
      rescue SyntaxError, StandardError
        {}
      end
```

File: lib/tmuxinator/cli.rb [hunk 5]
```diff
@@ -226,6 +227,44 @@ def show_continuation_prompt
...
+      def append_in_session(options, name = nil)
+        has_tmux_session = nil
+        if should_append?(options)
+          has_tmux_session = tmux_session_exists
+          if !has_tmux_session
+            say "Creating session '#{name}'"
+          end
+        end
+        has_tmux_session
+      end
```
Comment (noahfrederick) [inline lib/tmuxinator/cli.rb:249]: This method is somewhat hard to follow because it's responsible for two things: (a) deciding whether a new session is created *and* (b) printing a message with the new session name. Given (a), it should be named `append_in_session?` or something like that, and always return a Boolean (right now it returns `true`, `false` or `nil`—the rest of the code doesn't appear to distinguish between the `false` and `nil` possibilities). But since it also has a side effect of printing a message and takes this extra `name` parameter, that wouldn't be right either. These things don't seem to belong in the same method.

In any case, it could be simplified to:

```suggestion
      def append_in_session(options, name = nil)
        if should_append?(options) && !tmux_session_exists
          say "Creating session '#{name}'"
          true
        end
      end
```

File: lib/tmuxinator/project.rb (hunk 2 above)
Comment (noahfrederick) [inline lib/tmuxinator/project.rb:223]: Tell tmux to only print the window indices:

```suggestion
        base = 1 + `tmux list-windows -F '#I'`.split.last.to_i
```

The output is like so:

```
1
2
3
```

File: lib/tmuxinator/window.rb [hunk 6]
```diff
@@ -108,7 +108,8 @@ def tmux_window_command_prefix
     end
 
     def tmux_window_name_option
-      name ? "-n #{name}" : ""
+      project_name = project.in_current_session ? "#{project.full_name}:" : ""
+      name ? "-n #{project_name}#{name}" : ""
```
Comment (noahfrederick) [inline lib/tmuxinator/window.rb:112]: I would prefer to drop this change, i.e., for the window to always be named without the prefix.

**Review by noahfrederick — COMMENTED** [review]
Looking forward to this feature. I left a drive-by review of the code changes.

**Comment by flovilmart** [comment]
Thanks @noahfrederick will have a look later

File: lib/tmuxinator/config.rb [hunk 7]
```diff
@@ -56,6 +60,12 @@ def default
         "#{directory}/default.yml"
       end
 
+      def options
+        YAML.safe_load(File.read(options_file), [], [], true) || {}
```
Comment (ethagnawl) [inline lib/tmuxinator/config.rb:64]: @flovilmart Would you mind using the new safe_load method signature (sans positional parameters two, three and four)? As-is, this will cause errors in some environments because of changes to Psych and libyaml.

It would look something like this: 

`YAML.safe_load(File.read(options_file), aliases: true)`

See [this](https://github.com/tmuxinator/tmuxinator/pull/823/files) PR for more context.

**Comment by megalithic** [comment]
@flovilmart / @ethagnawl -- just pinging those that have worked on this. How are we feeling about your concerns being addressed and getting this merged?

**Comment by flovilmart** [comment]
@megalithic I haven't had the chance to actually finish this one; yeah I know, that's kinda pathetic after all this time 🤣

**Comment by ethagnawl** [comment]
@flovilmart I know your comment was in jest, but it's not pathetic at all. Your contributions are being provided at the expense of whatever else you might be doing with your life and are greatly appreciated.

**Comment by megalithic** [comment]
@ethagnawl @flovilmart 💯 agreed! OSS is a gift! i'm grateful for each of you that contribute/create such amazing tools.

**Comment by flovilmart** [comment]
Haha! Thanks y’all! very appreciated. I need to find some time for this. I see this is pretty requested.

**Comment by ChrisYin** [comment]
I'm wondering when this patch can be merged? Looks like it has a long history from 2018.

**Comment by martin-braun** [comment]
@flovilmart I know this issue probably hunts you in your dreams already, is there still any chance that you find the time to finish this?

**Comment by akofink** [comment]
I attempted to rebase this with `tmuxinator:master`, but I accidentally pushed the latest master, which auto-closed the PR. I have an updated branch that I will use to re-open a new PR, since I can no longer push to this original branch. Apologies!
